uxarray/core/api.py

```diff
@@ -0,0 +1,263 @@
+
+    Parameters
+    ----------
+
+    grid_filename_or_obj : string, xarray.Dataset, ndarray, list, tuple, required
+        String or Path object as a path to a netCDF file or an OpenDAP URL that
+        stores the unstructured grid topology/definition. It is read similar to
+        ``filename_or_obj`` in ``xarray.open_dataset``. Otherwise, either
+        ``xr.DataArray``, ``np.ndarray``, ``list``, or ``tuple`` as a vertices
+        object to define the grid.
+
+    islatlon : bool, optional
+            Specify if the grid is lat/lon based
+
+    isconcave: bool, optional
+        Specify if this grid has concave elements (internal checks for this are possible)
+
+    gridspec: str, optional
+        Specifies gridspec
+
+    vertices: bool, optional
+        Whether to create grid from vertices
+
+    source_grid: str, optional
+        Path or URL to the source grid file. For diagnostic/reporting purposes only.
+
+    use_dual: bool, optional
+            Specify whether to use the primal (use_dual=False) or dual (use_dual=True) mesh if the file type is mpas
+
+    **kwargs : Dict[str, Any]
+        Additional arguments passed on to ``xarray.open_dataset``. Refer to the
+        [xarray
+        docs](https://xarray.pydata.org/en/stable/generated/xarray.open_dataset.html)
+        for accepted keyword arguments.
+
+    Returns
+    -------
+
+    object : uxarray.Grid
+        Dataset with the unstructured grid.
+
+    Examples
+    --------
+
+    Open dataset with a grid topology file
+
+    >>> import uxarray as ux
+    >>> uxgrid = ux.open_grid("grid_filename.g")
+    """
+
+    ## Grid definition
+    if isinstance(grid_filename_or_obj,
+                  (list, tuple, np.ndarray, xr.DataArray)):
+        uxgrid = Grid(grid_filename_or_obj,
+                      gridspec=gridspec,
+                      vertices=vertices,
+                      islatlon=islatlon,
+                      isconcave=isconcave,
+                      source_grid=str(grid_filename_or_obj),
+                      use_dual=use_dual)
+    else:
+        grid_ds = xr.open_dataset(grid_filename_or_obj,
+                                  decode_times=False,
+                                  **kwargs)  # type: ignore
+
+        uxgrid = Grid(grid_ds,
+                      gridspec=gridspec,
+                      vertices=vertices,
+                      islatlon=islatlon,
+                      isconcave=isconcave,
+                      source_grid=str(grid_filename_or_obj),
+                      use_dual=use_dual)
+
+    return uxgrid
+
+
+def open_dataset(grid_filename_or_obj: str,
+                 filename_or_obj: str,
+                 gridspec: Optional[str] = None,
+                 vertices: Optional[list] = None,
+                 islatlon: Optional[bool] = False,
+                 isconcave: Optional[bool] = False,
+                 use_dual: Optional[bool] = False,
+                 **kwargs: Dict[str, Any]) -> xr.Dataset:
```

Comment:
Should this be `xr.Dataset` or `ux.UxDataset` ?

Reply:
Good catch; done!